src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,1990 @@
+    try System.getProperty("scala.collection.immutable.Vector.defaultApplyPreferredMaxLength",
+      "250").toInt
+    catch {
+      case _: SecurityException => 250
+    }
+
+  private val emptyIterator = new NewVectorIterator(Vector0, 0, 0)
+}
+
+
+/** Vector is a general-purpose, immutable data structure.  It provides random access and updates
+  * in O(log n) time, as well as very fast append/prepend/tail/init (amortized O(1), worst case O(log n)).
+  * Because vectors strike a good balance between fast random selections and fast random functional updates,
+  * they are currently the default implementation of immutable indexed sequences.
+  *
+  * Vectors are implemented by radix-balanced finger trees of width 32. There is a separate subclass
+  * for each level (0 to 6, with 0 being the empty vector and 6 a tree with a maximum width of 64 at the
+  * top level).
+  *
+  * Tree balancing:
+  * - Only the first dimension of an array may have a size < WIDTH
+  * - In a `data` (central) array the first dimension may be up to WIDTH-2 long, in `prefix1` and `suffix1` up
+  *   to WIDTH, and in other `prefix` and `suffix` arrays up to WIDTH-1
+  * - `prefix1` and `suffix1` are never empty
+  * - Balancing does not cross the main data array (i.e. prepending never touches the suffix and appending never touches
+  *   the prefix). The level is increased/decreased when the affected side plus main data is already full/empty
+  * - All arrays are left-aligned and truncated
+  *
+  * In addition to the data slices (`prefix1`, `prefix2`, ..., `dataN`, ..., `suffix2`, `suffix1`) we store a running
+  * count of elements after each prefix for more efficient indexing without having to dereference all prefix arrays.
+  */
+sealed abstract class Vector[+A] private[immutable] (private[immutable] final val prefix1: Arr1)
+  extends AbstractSeq[A]
+    with IndexedSeq[A]
+    with IndexedSeqOps[A, Vector, Vector[A]]
+    with StrictOptimizedSeqOps[A, Vector, Vector[A]]
+    with IterableFactoryDefaults[A, Vector]
+    with DefaultSerializable {
+
+  override def iterableFactory: SeqFactory[Vector] = Vector
+
+  protected[this] def tinyAppendLimit: Int = 4 + vectorSliceCount
+
+  override final def length: Int =
+    if(this.isInstanceOf[BigVector[_]]) this.asInstanceOf[BigVector[_]].length0
+    else prefix1.length
+
+  override final def iterator: Iterator[A] =
+    if(this.isInstanceOf[Vector0.type]) Vector.emptyIterator
+    else new NewVectorIterator(this, length, vectorSliceCount)
+
+  // Dummy overrides to refine result types for binary compatibility:
+  override def updated[B >: A](index: Int, elem: B): Vector[B] = super.updated(index, elem)
+  override def appended[B >: A](elem: B): Vector[B] = super.appended(elem)
+  override def prepended[B >: A](elem: B): Vector[B] = super.prepended(elem)
+  override def prependedAll[B >: A](prefix: collection.IterableOnce[B]): Vector[B] = super.prependedAll(prefix)
```

Comment:
It took me a while to bisect this and run benchmarks to be reasonably sure about this (sbt-cleaning everything is important; this doesn't interact well with incremental compilation) but these seemingly unrelated overrides cause a performance regression in `apply`. The results are actually a bit worse than what we had before removing the bounds checks. From this commit:

```
[info] Benchmark                           (size)  Mode  Cnt        Score       Error  Units
[info] VectorBenchmark2.nvApplySequential       1  avgt   20  6484281.171 ± 14144.596  ns/op
[info] VectorBenchmark2.nvApplySequential      10  avgt   20  6558224.871 ± 20183.258  ns/op
[info] VectorBenchmark2.nvApplySequential     100  avgt   20  7466587.037 ± 22809.037  ns/op
[info] VectorBenchmark2.nvApplySequential    1000  avgt   20  7494620.869 ± 25378.361  ns/op
[info] VectorBenchmark2.nvApplySequential   10000  avgt   20  8080951.377 ± 30335.256  ns/op
[info] VectorBenchmark2.nvApplySequential   50000  avgt   20  8257776.513 ± 12035.982  ns/op
```

With the overrides commented out:

```
[info] Benchmark                           (size)  Mode  Cnt        Score        Error  Units
[info] VectorBenchmark2.nvApplySequential       1  avgt   20  6020055.991 ± 125419.199  ns/op
[info] VectorBenchmark2.nvApplySequential      10  avgt   20  6098478.809 ± 100782.899  ns/op
[info] VectorBenchmark2.nvApplySequential     100  avgt   20  7097008.732 ±  54725.455  ns/op
[info] VectorBenchmark2.nvApplySequential    1000  avgt   20  7072626.835 ±  48439.117  ns/op
[info] VectorBenchmark2.nvApplySequential   10000  avgt   20  7811258.432 ±  50296.677  ns/op
[info] VectorBenchmark2.nvApplySequential   50000  avgt   20  8012436.017 ±   9179.761  ns/op
```

Reply:
This is an interesting one. It appears to be caused simply by the number of methods. Removing any one of the overrides makes it fast again and so does removing some other method. Fortunately `tinyAppendLimit` is completely useless and can be removed without repercussions.

Reply:
The `Vector` class also gained a bunch of helper methods (`mapElems1` ff) which don't really need to be class members. They are not used anywhere else so I put them directly into the class but if keeping the method count down is preferable they can easily be moved back into `VectorStatics`.